### DISCUSSION TITLE
create feed document

### DISCUSSION THREAD

**Discussion by esbezr** [saleweaver/python-amazon-sp-api#864] [Q&A] [answered]
I am coming up with an error for running the following code

Feeds().create_feed_document(file = "myfile.txt", content_type = "text/plain; charset=utf-8")

This is the error im getting which is believe is in reference to feed.py line 172 in create_feed_document, upload_data = file.read()
AttributeError: 'str' object has no attribute 'read'

**Comment by denisneuf** [comment]
Bro, the content_type is to specify which type of document you are uploading, is not a part of the api …. If you are using txt you need “text/plain”, if you use xml you need use "application/xml”,…etc

You can see a list of the possible values here:

https://www.geeksforgeeks.org/http-headers-content-type/

"

**Reply by esbezr** [reply]
I appreciate the answer. I had figured out that part and now im stuck on a different part. Have you ever used feeds to create a feed document?

**Reply by nordinho** [reply]
What do you need help with now?

**Reply by esbezr** [reply]
Im having trouble making a feed request. 

Feeds().create_feed_document(file = "myfile.txt", content_type = "text/plain; charset=utf-8")

This doesnt work and im not sure what im doing wrong. It returns the error: AttributeError: 'str' object has no attribute 'read'
When it gives this error its talking about the sp-api python script feed.py on line 172. On line 172 it says: upload_data = file.read(). I dont know what i am doing wrong but if you could help it would be appreciated.

**Reply by nordinho** [reply]
You need to turn it into a BytesIO format, try this:

from io import BytesIO
feed = BytesIO()
file = str.encode(file)
feed.write(file)
feed.seek(0)

Try this code and write file = feed, you may need to change type to 'text/xml'

**Reply by esbezr** [reply]
I appreciate the help. This worked great. Its now just failing in amazon. When i download the error report is says my sku field is missing from the text file i uploaded. 

These are the errors im getting. The template im pulling from matches my txt file so do i need change the content type to make this work or something?
90012	Fatal	The required sku field is missing from the file. To correct this error, download the template again and use the new copy, or insert the required field into your existing file.
90061	Warning	The myfile.txt field heading is invalid, so values in this field were ignored. To correct this error, download the template again and use the new copy, or fix the field heading in your existing file.

**Reply by GRBurst** [reply]
Is this solved for you? Please post your solution for further reference if you figured out the reason it failed on your side.
Otherwise, please provide a full minimal example.

**Comment by esbezr** [comment, marked as answer]
Solution:

import os
from sp_api.api import Feeds
from sp_api.base import Marketplaces
from io import BytesIO
import sys

feed = BytesIO()
file = str.encode("myfile.txt")
feed.write(file)
feed.seek(0)
        
res = Feeds(marketplace = Marketplaces.US).create_feed_document(file = feed, content_type = "text/html; charset=utf-8")
FDI = res.payload['feedDocumentId']
cre = Feeds(marketplace = Marketplaces.US).create_feed('POST_FLAT_FILE_INVLOADER_DATA', FDI)

**Reply by GRBurst** [reply]
ty 🙏
